packages/repository/src/type-resolver.ts

```diff
@@ -10,7 +10,10 @@ import {Class} from './common-types';
  * Intended to be used for cases when the JS engine is unable to fully define
  * a given type (require() loops).
  */
-export type TypeResolver<T extends Object> = () => Class<T>;
+export type TypeResolver<
+  Type extends Object,
+  StaticMembers = Function
```

Comment:
Can you clarify what `StaticMembers = Function` does? (sorry I havent seen it before)

Reply:
Let's say we have the following resolver:

```ts
const resolver: TypeResolver<Entity> = () => Customer;
const type = resolver();
```

Before my change, `type` was set to `Class<Entity>`, which is defined as a constructor function with arbitrary static properties.

```ts
export interface Class<T> {
  // new MyClass(...args) ==> T
  new (...args: any[]): T;
  // Other static properties/operations
  [property: string]: any;
}
```

These arbitrary static properties are effectively disabling compiler type checks and preventing VSCode from showing hints for static properties.

Consider the following snippet:

```ts
type.modelName
```

When `type` is just `Class<Entity>`, `modelName` is treated as an arbitrary indexer property of type `any`.

This pull request is making two changes:
1. Adding a `StaticMembers` template argument allowing users of `TypeProvider` to specify what static properties are expected on the resolved type. For example:

    ```ts
    const resolver: TypeResolver<Entity, typeof Entity> = () => Customer;
    const type = resolver();
    type.modelName 
    // ^-- TypeScript knows this is a `string` property
    // and VSCode shows a hint
    ```

2. Providing a default value for the new `StaticMembers` template argument. We use `Function` as the default value, because `Function` is the common denominator of all types (every class is a Function). As a result, the value returned by a type provider has only few static properties defined, e.g. `name: string`.

@b-admike Does it make sense now? I am going to capture this information in a code comment for future readers.

Reply:
What is also important: before my change, you could treat `modelName` as a function, write `model.modelName()`, the compiler will not complain and you will discover the mistake later at runtime. With my change in place, the compiler is able to type-check known static properties and point out that `modelName` is a string that cannot be invoked like a function.